core/modules/qproc/IndexMap.cc

```diff
@@ -0,0 +1,252 @@
+                scv.insert(scv.end(), area.begin(), area.end());
+            } else {
+                throw Bug("Null region ptr in IndexMap");
+            }
+        }
+        return scv;
+    }
+
+    inline SubChunksVector getCoverage(Region const& r) {
+        return _chunker->getSubChunksIntersecting(r);
+    }
+    ChunkSpecVector getAllChunks() const {
+        IntVector allChunks = _chunker->getAllChunks();
+        ChunkSpecVector csv;
+        csv.reserve(allChunks.size());
+        for(IntVector::const_iterator i=allChunks.begin(), e=allChunks.end();
+            i != e; ++i) {
+            csv.push_back(ChunkSpec(*i, _chunker->getAllSubChunks(*i)));
+        }
+        return csv;
+    }
+private:
+    boost::shared_ptr<sg::Chunker> _chunker;
+};
+
+////////////////////////////////////////////////////////////////////////
+// IndexMap implementation
+////////////////////////////////////////////////////////////////////////
+IndexMap::IndexMap(css::StripingParams const& sp,
+                   boost::shared_ptr<SecondaryIndex> si)
+    : _pm(boost::make_shared<PartitioningMap>(sp)),
+      _si(si) {
+
+}
+
+ChunkSpecVector IndexMap::getAll() {
+    return _pm->getAllChunks();
+}
+
+ChunkSpecVector IndexMap::getIntersect(query::ConstraintVector const& cv) {
+    // Index lookups
+    query::ConstraintVector indexConstraints;
+    //    std::copy_if(cv.begin(), cv.end(), std::back_inserter(indexConstraints),
+    //                 isIndex);
+    // copy_if not available before c++11
+    std::copy(cv.begin(), cv.end(), std::back_inserter(indexConstraints));
+    query::ConstraintVector::const_iterator newEnd;
+    newEnd = std::remove_if(indexConstraints.begin(), indexConstraints.end(),
+                            isNotIndex);
+    indexConstraints.resize(newEnd - indexConstraints.begin());
+    ChunkSpecVector indexSpecs;
+    if(!indexConstraints.empty()) {
+        if(!_si) {
+            throw Bug("Invalid SecondaryIndex in IndexMap. Check IndexMap(...)");
+        }
+        indexSpecs = _si->lookup(indexConstraints);
+    }
+
+    // Spatial area lookups
+    if(indexConstraints.size() == cv.size()) {
+        if(indexConstraints.empty()) { // No constraints -> full-sky
+            return _pm->getAllChunks();
+        } else {
+            // No spatial constraints?
+            return indexSpecs;
+        }
+    } else {
+        // Index and spatial lookup are supported in AND format only right now.
+        // If both exist, compute the AND.
+        RegionPtrVector rv;
+        std::transform(cv.begin(), cv.end(), std::back_inserter(rv), getRegion);
+
+        SubChunksVector scv = _pm->getIntersect(rv);
+
+        ChunkSpecVector csv;
+        std::transform(scv.begin(), scv.end(),
+                       std::back_inserter(csv), convertSgSubChunks);
+        if(!indexSpecs.empty()) {
+            intersectSorted(csv, indexSpecs); // performs "AND"
+        }
+        return csv;
+    }
+}
```

Comment:
I feel like this would be a lot simpler and involve a lot less vector copying/creation etc... if you made SecondaryIndex just ignore non "sIndex" constraints, and ditched the use of `std::transform`, `std::copy`, etc...